packages/adapter-components/src/client/rate_limiter.ts

```diff
@@ -191,17 +246,72 @@ export class RateLimiter {
 
   /**
    * Wraps a given task with counters bookkeeping.
-   * @param task The task to wrapped with bookkeeping.
-   * @returns A function that returns a promise resolving to the result of the task
+   * @param task The task to be wrapped with bookkeeping.
+   * @returns A function that returns a promise resolving to the result of the task.
    */
-  private wrapWithBookKeeping<T>(task: () => Promise<T>): () => Promise<T> {
+  private wrapWithBookKeeping<T>(task: TaskType<T>, isRetry: boolean = false): TaskType<T> {
     return async (): Promise<T> => {
       this.internalCounters.pending -= 1
       this.internalCounters.running += 1
-      const res = await task()
-      this.internalCounters.running -= 1
-      this.internalCounters.done += 1
-      return res
+      if (isRetry) {
+        this.internalCounters.retries += 1
+      }
+      let res
+      try {
+        res = await task()
+        this.internalCounters.succeeded += 1
+        return res
+      } catch (e) {
+        this.internalCounters.failed += 1
+        throw e
```

Comment:
nit - instead of having `wrapWithPauseOnFail` method:

```suggestion
        this.internalCounters.failed += 1
        if (this.internalOptions.pauseDuringRetryDelay) {
          this.pause(delay)        
        }
        throw e
```

I know that it's not related to "book keeping", but eventually having less wrapping functions makes the code much easier to follow.